README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# simK <img src="man/figures/logo.png" height="150" align="right"/>

[![Project Status: Active – The project has reached a st able, usable
state and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)

**License**: Creative Commons Attribution
[CC-BY](https://creativecommons.org/licenses/by/2.0/). Feel free to
share and adapt, but don’t forget to credit the author. :sunglasses:

:dart: Functions to procedurally generate synthetic data in R for kidney
transplant simulations. You can see `simK` in action,
[here](https://simk.netlify.app/).

`simK` allows to generate data with clinical and demographic information
for a pool of simulated cadaveric donors and simulated wait listed
candidates for kidney transplantation.

Data generated with `simK` are particularly useful on
[KARS](https://balima.shinyapps.io/kars/).

> :warning: This package is not a medical device and <ins>should not be
> used for making clinical decisions</ins>.

## Instalation

### Development version

The development version can be installed from GitHub, if you want all
the latest features, together with all the latest bugs and errors. You
have been warned :exclamation:

    # install from GitHub
    devtools::install_github("balima78/simK")

## Main functions

This package has 3 main functions, with them we can generate simulated
data for a pool of donors, a set of kidney transplant candidates and the
respective HLA-antibodies for those patients HLA sensitized.

A detailed description of these functions usage, is available from
[here](https://simk.netlify.app/articles/simk.html).

:warning: All these functions rely on HLA typing at intermediate
resolution as described at [Lima *et al*,
2013](https://12f11c1f-960a-f627-594d-b8ce276384f7.filesusr.com/ugd/3e838e_dc548dede99a4db5869c3d2c20c2d16f.pdf?index=true).
:warning:


NEWS.md
# simK 0.1.1

* changed functions names to snake case

* added require(magrittr) to functions using %>% 

# simK 0.1.0

* first release version


R/data.R
#' HLA typing.
#'
#' A dataset containing HLA typing from 37993 unrelated voluntary bone marrow donors from Nothern Portugal, recruited between 2010 and 2011.
#'
#' @format A data frame with 37993 rows and 9 variables:
#' \describe{
#'   \item{ID}{Identifier}
#'   \item{A1}{HLA-A allele 1}
#'   \item{A2}{HLA-A allele 2}
#'   \item{B1}{HLA-B allele 1}
#'   \item{B2}{HLA-B allele 2}
#'   \item{C1}{HLA-C allele 1}
#'   \item{C2}{HLA-C allele 2}
#'   \item{DRB11}{HLA-DRB1 allele 1}
#'   \item{DRB12}{HLA-DRB1 allele 2}
#'   ...
#' }
#' @source \url{https://www.slideshare.net/balima78/lima-2013}
"hla"


#' A pool of 10.0000 donors.
#'
#' A synthetic dataset with information for 10.000 donors, generated with `donors.df()` and `set.seed(3)` .
#'
#' @format A data frame with 37993 rows and 9 variables:
#' \describe{
#'   \item{ID}{Identifier}
#'   \item{bg}{ABO blood group}
#'   \item{A1}{HLA-A allele 1}
#'   \item{A2}{HLA-A allele 2}
#'   \item{B1}{HLA-B allele 1}
#'   \item{B2}{HLA-B allele 2}
#'   \item{DR1}{HLA-DRB1 allele 1}
#'   \item{DR2}{HLA-DRB1 allele 2}
#'   \item{age}{age in years}
#'   ...
#' }
#' @source \url{https://www.slideshare.net/balima78/lima-2013}
"D10K"


#' A vector with HLA-A.
#'
#' A character vector with antigens for HLA-A locus obtained from \code{hla} dataset.
#'
#' @format A character vector with length 20.
"agA"

#' A vector with HLA-B.
#'
#' A character vector with antigens for HLA-B locus obtained from \code{hla} dataset.
#'
#' @format A character vector with length 34.
"agB"

#' A vector with HLA-DR.
#'
#' A character vector with antigens for HLA-DR locus obtained from \code{hla} dataset.
#'
#' @format A character vector with length 13.
"agDR"


R/histo.R
#' number of HLA mismatchs
#'
#' @description Computes the number of HLA mismatchs between one donor and one candidate
#' @param dA donor's HLA-A typing
#' @param dB donor's HLA-B typing
#' @param dDR donor's HLA-DR typing
#' @param cA candidate's HLA-A typing
#' @param cB candidate's HLA-B typing
#' @param cDR candidate's HLA-DR typing
#' @return mmA number of HLA-A mismatchs between \code{dA} and \code{cA};
#' mmB number of HLA-B mismatchs between \code{dB} and \code{cB};
#' mmDR number of HLA-DR mismatchs between \code{dDR} and \code{cDR};
#' and mmHLA as the sum of mmA + mmB + mmDR
#' @examples
#' mmHLA(dA = c('1','2'), dB = c('5','7'), dDR = c('1','4'),
#' cA = c('1','2'), cB = c('03','15'), cDR = c('04','07'))
#' @export
#' @concept histocompatibility
mmHLA <- function(dA = c('1','2'), dB = c('5','7'), dDR = c('1','4'),
                  cA = c('1','2'), cB = c('3','15'), cDR = c('4','7')){

  mmA = NULL
  mmB = NULL
  mmDR = NULL

  # verify function parameters
  if(!is.character(dA)){stop("donor's HLA-A typing is not valid!\n")}
  if(!is.character(dB)){stop("donor's HLA-B typing is not valid!\n")}
  if(!is.character(dDR)){stop("donor's HLA-DR typing is not valid!\n")}
  if(!is.character(cA)){stop("candidate's HLA-A typing is not valid!\n")}
  if(!is.character(cB)){stop("candidate's HLA-B typing is not valid!\n")}
  if(!is.character(cDR)){stop("candidate's HLA-DR typing is not valid!\n")}

  # compute missmatches
  mmA<-ifelse((dA[1] %in% cA & dA[2] %in% cA) | (dA[1] %in% cA & (is.na(dA[2]) | dA[2] == "")), 0,
              ifelse(dA[1] %in% cA | dA[2] %in% cA, 1,
                     ifelse(!dA[1] %in% cA & (is.na(dA[2]) | dA[2] == ""), 1,
                            ifelse(dA[1] == dA[2], 1,2))))

  mmB<-ifelse((dB[1] %in% cB & dB[2] %in% cB) | (dB[1] %in% cB & (is.na(dB[2]) | dB[2] == "")), 0,
              ifelse(dB[1] %in% cB | dB[2] %in% cB, 1,
                     ifelse(!dB[1] %in% cB & (is.na(dB[2]) | dB[2] == ""), 1,
                            ifelse(dB[1] == dB[2], 1,2))))

  mmDR<-ifelse((dDR[1] %in% cDR & dDR[2] %in% cDR) | (dDR[1] %in% cDR & (is.na(dDR[2]) | dDR[2] == "")), 0,
               ifelse(dDR[1] %in% cDR | dDR[2] %in% cDR, 1,
                      ifelse(!dDR[1] %in% cDR & (is.na(dDR[2]) | dDR[2] == ""), 1,
                             ifelse(dDR[1] == dDR[2],1,2))))

  # resume results
  mmHLA = mmA + mmB + mmDR
  mm = c(mmA,mmB,mmDR,mmHLA)
  names(mm) <- c("mmA","mmB","mmDR","mmHLA")

  return(mm)
}

#' Matchability from D10K
#'
#' @description Computes the number donors on dataset D10K that are a match to
#' a given transplant candidate. A sample of D10K is selected according to
#' cPRA value, and donors ABO identical and HLA mismatch level 1 or 2
#' (0 DR or (1 DR and 0 B)) are filtered.
#' @param cABO A character from 'A', 'B', 'AB', 'O'
#' @param cPRA candidate's cPRA value
#' @param cA candidate's HLA-A typing
#' @param cB candidate's HLA-B typing
#' @param cDR candidate's HLA-DR typing
#' @param n_seed a numeric seed that will be used for random number generation.
#' @return Match Score measure of how difficult it is to match a patient with a organ donor. A score from 1 (easy to match) to 10 (difficult to match).
#' @examples
#' matchability(cABO = 'A', cPRA = 85,
#' cA = c('2','29'), cB = c('7','15'), cDR = c('4','7'),
#' n_seed = 3)
#' @export
#' @concept histocompatibility
matchability <- function(cABO = 'A', cPRA = 85,
                         cA = c('2','29'), cB = c('7','15'), cDR = c('4','7'),
                         n_seed = 3){
  if(!cABO %in% c('A','AB','B','O')){stop("Blood group is not valid! Valid options: 'A','AB','B','O'")}

  require(magrittr)

  set.seed(n_seed)

  n1 <- (100-cPRA)*100

  n.donors <- dplyr::sample_n(D10K, size = n1) %>%
    dplyr::filter(bg == cABO) %>%
    dplyr::mutate_if(is.numeric, as.character) %>%
    dplyr::rowwise() %>%
    dplyr::mutate(mmB = mmHLA(dA = c(A1,A2), dB = c(B1,B2), dDR = c(DR1,DR2),
                              cA = cA, cB = cB, cDR = cDR)['mmB'],
                  mmDR = mmHLA(dA = c(A1,A2), dB = c(B1,B2), dDR = c(DR1,DR2),
                               cA = cA, cB = cB, cDR = cDR)['mmDR'],
                  level12 = mmDR == 0 | (mmB == 0 & mmDR == 1)) %>%
    dplyr::ungroup() %>%
    dplyr::filter(level12) %>% nrow()

  return(n.donors)

}

#' virtual PRA
#'
#' @description Computes virtual PRA (vPRA) form HLA-A, -B, -DR loci.
#' @param abs A character vector with HLA antibodies.
#' @param donors A dataframe with HLA typing for a pool of donors.
#' @return a percentual value for vPRA
#' @examples
#' vpra(abs = c('A1','A2','B5','DR4'), donors = D10K)
#' @export
#' @concept histocompatibility
vpra <- function(abs = c('A1','A2','B5','DR4'), donors = D10K){

  require(magrittr)

  n <- nrow(D10K)

  na <- donors %>%
    dplyr::mutate_at(dplyr::vars(A1,A2), function(x) paste0('A',x)) %>%
    dplyr::mutate_at(dplyr::vars(B1,B2), function(x) paste0('B',x)) %>%
    dplyr::mutate_at(dplyr::vars(DR1,DR2), function(x) paste0('DR',x)) %>%
    dplyr::filter(A1 %in% abs | A2 %in% abs |
                    B1 %in% abs | B2 %in% abs |
                    DR1 %in% abs | DR2 %in% abs) %>%
    nrow()

  res <- na/n * 100

  return(res)

}

#' samples HLA antibodies
#'
#' @description creates a sample of HLA antibodies (abs) for a given candidate
#' according with a cPRA value.
#' @param cA candidate's HLA-A typing
#' @param cB candidate's HLA-B typing
#' @param cDR candidate's HLA-DR typing
#' @param cPRA candidate's cPRA value
#' @param n_seed a numeric seed that will be used for random number generation.
#' @return a character vector with HLA abs.
#' @examples
#' antbs(cA = c('2','29'), cB = c('7','15'), cDR = c('4','7'), cPRA = 85, n_seed = 3)
#' @export
#' @concept histocompatibility
antbs <- function(cA = c('2','29'), cB = c('7','15'), cDR = c('4','7'),
                cPRA = 85, n_seed = 3){

  set.seed(n_seed)

  typing <- c(paste0('A',cA), paste0('B',cB), paste0('DR',cDR))

  valid.ags <- c(agA, agB, agDR)[!c(agA, agB, agDR) %in% typing]

  c = NULL
  if(cPRA > 0){
    for(i in 1:250){
      c[i] <- sample(valid.ags, 1, replace = F)

      if(vpra(c)>cPRA-3){ break }
    }
  }

  vpra <- vpra(c)

  list(cPRA = cPRA,
       vPRA = vpra,
       HLA = typing,
       Valid.Ags = valid.ags,
       Abs = c)

}


rascunho.R
usethis::use_package("data.table")

usethis::use_testthat()

library(tidyverse)
library(data.table)

usethis::use_package("transplantr", "Imports")

# Edit one or more files below R/.
# document() (if you’ve made any changes that impact help files or NAMESPACE)
# load_all()
# Run some examples interactively.
# test() (or test_file())
# check()

devtools::document()
devtools::load_all()
devtools::test()

devtools::check()

hla <- read_csv2("~/2.PhD/HEADS/HLA37993.csv", col_types = 'cdddddddd')

usethis::use_data(D10K)

usethis::use_ccby_license()

usethis::use_news_md(open = rlang::is_interactive())

usethis::use_pkgdown()

pkgdown::build_site()
usethis::use_pkgdown_github_pages()

# https://rich-iannone.github.io/pointblank/articles/VALID-III.html

set.seed(3)
donors.df(uk = T)


library(transplantr)

kidney.donors

data(kidney.recipients)

ukkdri(age = 55, height = 180, htn = 1, sex = 'M', cmv = 1, gfr = 30, hdays = 100)
# https://cran.r-project.org/web/packages/transplantr/vignettes/kidney_risk_scores.html
# https://transplantr.txtools.net/articles/kidney_risk_scores.html
# https://cran.r-project.org/web/packages/transplantr/

library(tictoc)
tic()
candidates.df(n = 2000, uk=T)
toc()


# continuar a partir da pasta:
# D:\backup CNH bruno 2020-12-15\_CHN_\Trabalhos\CHN\Allocation 2017-09
# para criar ficheiro de Acs


library(tictoc)
tic()
candidates <- candidates.df(n=10, uk=T)
toc() # 5.5''

tic()
donors <- donors.df(n=10, uk=T)
toc() # 0.06''

tic()
abs <- Abs.df(candidates = candidates)
toc() # 4.65''

Abs.df(candidates = candidates.df(n=10),
       n.seed = 3)

library(hexSticker)
imgurl <- system.file("figures/kidneys.JPG", package="hexSticker")
s <- sticker(imgurl, package="simK", p_size=20, p_y = 1.65, p_color = 'black',
             s_x=1, s_y=1, s_width=.6,
             h_fill = 'white', h_color = 'red', h_size = 8,
             spotlight = T,
             url = 'https://txor.netlify.app/', u_size = 6,
             white_around_sticker = T,
        filename="images/simk.png")

candidates_df(n = 10,
              replace = TRUE,
              probs_abo = c(0.43, 0.03, 0.08, 0.46),
              probs_cpra = c(0.7, 0.1, 0.1, 0.1),
              lower = 18, upper = 75, mean = 45, sd = 15,
              prob_dm = 0.12,
              uk = TRUE,
              n_seed = 3)


Abs_df(candidates = candidates_df(n=10),
       n_seed = 3)



R/clinical.R
#' A data frame with HLA typing
#'
#' @description Returns a data frame with HLA-A, -B, -C and -DRB1 typing
#' @param n An integer to define the number of rows
#' @param replace A logical value for sampling with replacement
#' @return A data frame
#' @examples
#' hla_sample(n = 1000, replace = TRUE)
#' @export
#' @concept clinical_parameters
hla_sample <- function(n, replace){

  if(!is.numeric(n)){stop("`n` must be a single number!")}
  if(!is.logical(replace)){stop("`replace` must be logical (TRUE or FALSE)")}

  df <- dplyr::slice_sample(hla, n = n, replace = replace)

  return(df)
}

#' Builds a vector with ABO blood group
#'
#' @description Returns a vector with ABO blood groups according to user defined
#' frequencies
#' @param n An integer to define the length of the returned vector
#' @param probs A vector with the probabilities for blood group A, AB, B and O (in this order). The sum of the probabilities must be equal to one.
#' @return A vector length `n` with ABO blood groups
#' @examples
#' abo(n = 100, probs = c(0.4658, 0.0343, 0.077, 0.4229))
#' @export
abo <- function(n = 100, probs = c(0.4658, 0.0343, 0.077, 0.4229)){

  if(!is.numeric(n) | n < 1){stop("`n` must be a single number!")}
  if(round(sum(probs)) != 1){stop("`probs` do not sum 1!")}

  abo <- c('A','AB','B','O')
  sample(x=abo, size = n, replace = TRUE, prob = probs)

}

#' Gives a eGFR by age
#'
#' @description Returns a value for the Estimated Glomerular Filtration Rate
#' (eGFR) by age as described by https://www.kidney.org/atoz/content/gfr.
#' @param age An integer for age (values between 1 and 99).
#' @return A value from a normal distribution.
#' @examples
#' aGFR(age = 43)
#' @export
#' @concept clinical_parameters
aGFR <- function(age = 43){
  if(!is.numeric(age) | age < 1 | age > 99){stop("`age` must be between 1 and 99!")}

  res <- ifelse(age < 30, truncnorm::rtruncnorm(n = 1, a=106, b=122, mean = 116, sd = 10),
                ifelse(age < 40, truncnorm::rtruncnorm(n = 1, a=97, b=117, mean = 107, sd = 10),
                       ifelse(age < 50, truncnorm::rtruncnorm(n = 1, a=89, b=109, mean = 99, sd = 10),
                              ifelse(age < 60, truncnorm::rtruncnorm(n = 1, a=83, b=103, mean = 93, sd = 10),
                                     ifelse(age < 70, truncnorm::rtruncnorm(n = 1, a=75, b=95, mean = 85, sd = 10),
                                            truncnorm::rtruncnorm(n = 1, a=65, b=85, mean = 75, sd = 10))))))

  return(res)

}

#' Builds a vector with cPRA (percentage)
#'
#' @description Returns a vector with cPRA percentages according to user
#' defined frequencies
#' @param n An integer to define the length of the returned vector
#' @param probs A vector with the probabilities for cPRA groups 0%, 1%-50%, 51%-84%, 85%-100% (in this order). The sum of the probabilities must be equal to one.
#' @return A vector length `n` with cPRA percentages
#' @examples
#' cpra(n = 100, probs = c(0.7, 0.1, 0.1, 0.1))
#' @export
#' @concept clinical_parameters
cpra <- function(n = 100, probs = c(0.7, 0.1, 0.1, 0.1)){

  if(!is.numeric(n) | n < 1){stop("`n` must be a single number!")}
  if(round(sum(probs)) != 1){stop("`probs` do not sum 1!")}
  if(length(probs) != 4){stop("`probs` must be a vector with length 4!")}

  n4 <- round(n*probs[4])
  n3 <- round(n*probs[3])
  n2 <- round(n*probs[2])
  n1 <- n - (n2+n3+n4)

  v1 <- rep(0,n1)
  v2 <- extraDistr::rtpois(n = n2, lambda = 30, a = 1, b = 50)
  v3 <- extraDistr::rtpois(n = n3, lambda = 70, a = 51, b = 85)
  v4 <- extraDistr::rtpois(n = n4, lambda = 90, a = 86, b = 100)

  v <- c(v1,v2,v3,v4)

  return(sample(v))
}

#' Gives time on dialysis in months
#'
#' @description Returns a value for time on dialysis in months by blood group and cPRA.
#' @param hiper A logical value for hipersensitized patients (cPRA > 85%).
#' @param bg A character value for blood group.
#' @return A value from a normal distribution.
#' @examples
#' dial(hiper = TRUE, bg = 'O')
#' @export
#' @concept clinical_parameters
dial <- function(hiper = TRUE, bg = 'O'){
  if(!bg %in% c('A','AB','B','O')){stop("`bg` is not valid! valid blood group: 'A','AB','B','O'")}
  if(!is.logical(hiper)){stop("`hiper` must be a logical value!")}

  dial1 <- round(rnorm(1, 35,20))

  res <- ifelse(hiper == TRUE & bg == 'O', round(rnorm(1, 85,20)),
                ifelse(hiper == TRUE | bg == 'O', round(rnorm(1, 70,20)),
                       ifelse(dial1 < 0, 0, dial1)
                       ))

  return(res)

}





R/demographic.R
#' Builds a vector with truncated ages
#'
#' @description Returns a vector with ages with a normal distribution truncated by upper anf upper limits.
#' @param n An integer to define the length of the returned vector.
#' @param lower An integer for ages' lower limit.
#' @param upper An integer for ages' upper limit.
#' @param mean A value for mean's distribution.
#' @param sd A value for standar deviation's distribution.
#' @return A vector length `n` with normal distributed truncated ages.
#' @examples
#' ages(n = 100, lower=18, upper=75, mean = 55, sd = 15)
#' @export
ages <- function(n = 100, lower=18, upper=75, mean = 55, sd = 15){

  if(!is.numeric(n) | n < 1){stop("`n` must be a positive single number!\n")}
  if(!is.numeric(mean) | n < 1){stop("`mean` must be a positive single number!\n")}
  if(!is.numeric(sd) | n < 1){stop("`sd` must be a positive single number!\n")}
  if(!is.numeric(lower) | n < 1){stop("`lower` must be a positive single number!\n")}
  if(!is.numeric(upper) | n < 1){stop("`upper` must be a positive single number!\n")}

  round(
    truncnorm::rtruncnorm(n = n, a=lower, b=upper, mean = mean, sd = sd)
  )

}





R/sim_df.R
#' A data frame with donors caracheristics
#'
#' @description Returns a data frame for simulated donors clinical and
#' demographic caracheristics.
#' @param n An integer to define the number of rows
#' @param replace A logical value for sampling with replacement
#' @param probs A vector with the probabilities for blood group A, AB, B and O (in this order). The sum of the probabilities must be equal to one.
#' @param lower An integer for ages' lower limit.
#' @param upper An integer for ages' upper limit.
#' @param mean A value for mean's distribution.
#' @param sd A value for standar deviation's distribution.
#' @param uk A logical value, if TRUE is also computed the Donor Risk Index (DRI)
#' @param n_seed a numeric seed that will be used for random number generation.
#' @return A data frame with HLA typing, blood group and truncated ages for a simulated group of transplant donors.
#' @examples
#' donors_df(n = 10, replace = TRUE,
#' probs = c(0.4658, 0.0343, 0.077, 0.4229),
#' lower=18, upper=75, mean = 55, sd = 15,
#' uk = FALSE, n_seed = 3)
#' @export
#' @concept generate_data
donors_df <- function(n = 10, replace = TRUE,
                      probs = c(0.4658, 0.0343, 0.077, 0.4229),
                      lower=18, upper=75,
                      mean = 60, sd = 12,
                      uk = FALSE,
                      n_seed = 3){

  require(magrittr)

  set.seed(n_seed)

  df <- hla_sample(n = n, replace = replace)
  df$bg <- abo(n = n, probs = probs)
  df$age <- ages(n = n, lower=lower, upper=upper, mean = mean, sd = sd)
  df$ID <- paste0('D', 1:n)

  df <- df %>%
    dplyr::rename(DR1 = DRB11, DR2 = DRB12) %>%
    dplyr::select(ID, bg, A1, A2, B1, B2, DR1, DR2, age)

  if(uk == TRUE){
    # a column for height with N(165,20)
    df$height <- truncnorm::rtruncnorm(n = n, a=150, b=200, mean = 165, sd = 20)
    # # a column for hypertension with a mean frequency of 43%
    df$hypertension <- sample(c(1,0), size = n, replace = TRUE, prob = c(0.4, 0.6))
    # # a column for sex with a mean frequency of men (M) = 55%
    df$sex <- sample(c('M','F'), size = n, replace = TRUE, prob = c(0.55,0.45))
    # # a column for CMV+ with a mean frequency of 90%
    df$cmv <- sample(c(1,0), size = n, replace = TRUE, prob = c(0.9, 0.1))
    # # a column for nº of days stayed on Hospital with P(4)
    df$hospital_stay <- rpois(n = n, lambda = 4)

    df<-df %>%
      dplyr::mutate(GFR = purrr::map_dbl(age,
                                         ~aGFR(age = .x)),
             UKKDRI = transplantr::ukkdri(age = age, height = height,
                                          htn = hypertension, sex = sex,
                                          cmv = cmv, gfr = GFR,
                                          hdays = hospital_stay),
             DRI = transplantr::ukkdri_q(UKKDRI, prefix = TRUE)
             ) %>%
      dplyr::mutate_at(dplyr::vars(tidyselect::ends_with('1'), tidyselect::ends_with('2')), as.character) %>%
      dplyr::select(ID, bg, A1, A2, B1, B2, DR1, DR2, age, DRI)
  }

  return(df)
}

#' A data frame with donors characteristics
#'
#' @description Returns a data frame for simulated donors clinical and
#' demographic characteristics.
#' @param n An integer to define the number of rows
#' @param replace A logical value for sampling with replacement
#' @param probs_abo A vector with the probabilities for blood group A, AB, B and O (in this order). The sum of the probabilities must be equal to one.
#' @param probs_cpra A vector with the probabilities for cPRA groups 0%, 1%-50%, 51%-84%, 85%-100% (in this order). The sum of the probabilities must be equal to one.
#' @param lower An integer for ages' lower limit.
#' @param upper An integer for ages' upper limit.
#' @param mean A value for mean age's distribution.
#' @param sd A value for standard deviation age's distribution.
#' @param prob_dm A value for the probability of having Diabetes Mellitus
#' @param uk A logical value, if TRUE is also computed the Donor Risk Index (DRI)
#' @param n_seed a numeric seed that will be used for random number generation.
#' @return A data frame with HLA typing, blood group, truncated ages, time on dialysis (in months), cPRA, Tier, MS and RRI (those last 3, only when uk = TRUE) for a simulated group of transplant candidates.
#' @examples
#' candidates_df(n = 10, replace = TRUE,
#' probs_abo = c(0.43, 0.03, 0.08, 0.46), probs_cpra = c(0.7, 0.1, 0.1, 0.1),
#' lower=18, upper=75, mean = 45, sd = 15,
#' prob_dm = 0.12,
#' uk = FALSE, n_seed = 3)
#' @export
#' @concept generate_data
candidates_df <- function(n = 10, replace = TRUE,
                          probs_abo = c(0.43, 0.03, 0.08, 0.46),
                          probs_cpra = c(0.7, 0.1, 0.1, 0.1),
                          lower=18, upper=75,
                          mean = 45, sd = 15,
                          prob_dm = 0.12,
                          uk = FALSE,
                          n_seed = 3){

  require(magrittr)

  set.seed(n_seed)

  df <- hla_sample(n = n, replace = replace)
  df$bg <- abo(n = n, probs = probs_abo)
  df$cPRA <- cpra(n = n, probs = probs_cpra)
  df$age <- ages(n = n, lower=lower, upper=upper, mean = mean, sd = sd)
  df$ID <- paste0('K', 1:n)

  df <- df %>%
    dplyr::rename(DR1 = DRB11, DR2 = DRB12) %>%
    dplyr::select(ID, bg, A1, A2, B1, B2, DR1, DR2, age, cPRA) %>%
    dplyr::mutate(hiper = cPRA > 85,
                  dialysis = purrr::map2_dbl(.x = bg,
                                      .y = hiper,
                                      ~dial(hiper = .y, bg = .x)))
  if(uk == TRUE){
    dm1 <- prob_dm
    dm0 <- 1-prob_dm
    df$dm <- sample(c(0,1), size = n, replace = TRUE, prob = c(dm0,dm1))
    df<-df %>%
      dplyr::mutate(UKKRRI = transplantr::ukkrri(age = age,
                                                 dx = 1,
                                                 wait = dialysis * 30,
                                                 dm = dm),
                    RRI = transplantr::ukkrri_q(UKKRRI, prefix = T)
                    ) %>%
      dplyr::mutate_at(dplyr::vars(tidyselect::ends_with('1'),tidyselect::ends_with('2')), as.character) %>%
      dplyr::rowwise() %>%
      dplyr::mutate(ms = matchability(cABO = bg, cPRA = cPRA,
                                      cA = c(A1,A2), cB = c(B1,B2), cDR = c(DR1,DR2),
                                      n_seed = n_seed)
             ) %>%
      dplyr::ungroup()
    df$MS <- dplyr::ntile(dplyr::desc(df$ms), 10)
    df <- df %>%
      dplyr::mutate(Tier = ifelse(MS == 10 | cPRA == 100 | dialysis > 7*12, 'A', 'B')) %>%
      dplyr::select(ID, bg, A1, A2, B1, B2, DR1, DR2, age, dialysis, cPRA, Tier, MS, RRI)


    }

  return(df)
  }

#' A data frame with candidates' HLA antibodies
#'
#' @description Returns a data frame with transplant candidates' HLA antibodies
#' obtained according to theirs cPRA values and HLA typing.
#' @param candidates A dataframe with \code{ID}, HLA typing (column names: \code{A1}, \code{A2}, \code{B1}, \code{B2}, \code{DR1}, \code{DR2}) and cPRA value (column name: \code{cPRA}), for a group of transplant candidates.
#' @param n_seed a numeric seed that will be used for random number generation.
#' @return A data frame with \code{ID} and HLA antibodies \code{Abs}.
#' @examples
#' Abs_df(candidates = candidates_df(n=10), n_seed = 3)
#' @export
#' @concept generate_data
Abs_df <- function(candidates = candidates_df(n=10), n_seed = 3){

  require("magrittr")

  df <- candidates %>%
    dplyr::rowwise() %>%
    dplyr::mutate(abs = list(antbs(cA = c(A1,A2), cB = c(B1,B2), cDR = c(DR1,DR2),
                                    cPRA = cPRA,
                                    n_seed = n_seed)$Abs)) %>%
    dplyr::ungroup()

  df %>%
    dplyr::filter(cPRA > 0) %>%
    dplyr::select(ID, abs) %>%
    tidyr::unnest(abs)

}
